## Pagination and Downloading Data

LuxY is designed to handle pagination in alignment with the Lux API. To learn about how to work with pagination and download data from Lux, let's start by trying to find all the PeopleGroups that have the name "John Adams".

In [1]:
from luxy import PeopleGroups

In [2]:
pg = PeopleGroups().filter(name="John Adams").get()
# print the number of results
print("Number of results:", pg.num_results)
print("Number of pages:", pg.num_pages())

Number of results: 276
Number of pages: 14


Here, we can see our results. As of the creation of this notebook, there are 276 total results and 14 pages. We can use LuxY to iterate over those pages by using the `get_page_data_all` method. We can pass two arguments to this method: `start_page` and `end_page`. These arguments are optional, and if not provided, will default to 1 and the total number of pages, respectively.

In [15]:
pages = pg.get_page_data_all(start_page=1, end_page=3)

As we can see below, this returns a generator object. We can convert this to a list by using the `list` function.

In [16]:
pages

<generator object BaseLux.get_page_data_all at 0x111c21640>

In [17]:
pages = list(pages)

Let's take a look at the first page.

In [18]:
print(pages)

[{'@context': 'https://linked.art/ns/v1/search.json', 'id': 'https://lux.collections.yale.edu/api/search/agent?q=%7B%22AND%22%3A%5B%7B%22name%22%3A%22John%20Adams%22%7D%5D%7D&page=2&pageLength=20', 'type': 'OrderedCollectionPage', 'partOf': [{'id': 'https://lux.collections.yale.edu/api/search-estimate/agent?q=%7B%22AND%22%3A%5B%7B%22name%22%3A%22John%20Adams%22%7D%5D%7D', 'type': 'OrderedCollection', 'label': {'en': ['People & Groups']}, 'summary': {'en': ['Records representing individuals and organizations that match your search.']}, 'totalItems': 276}], 'orderedItems': [{'id': 'https://lux.collections.yale.edu/data/person/22e467d0-8f37-4fd8-a317-7eb096375ee0', 'type': 'Person'}, {'id': 'https://lux.collections.yale.edu/data/person/4c7f7739-f7be-4e0d-9c88-987ee8bbc3f8', 'type': 'Person'}, {'id': 'https://lux.collections.yale.edu/data/person/67e1c047-f6b7-412e-a000-c095333e649b', 'type': 'Person'}, {'id': 'https://lux.collections.yale.edu/data/person/bb5f8245-4d82-4d0f-8ce8-b8b81230ac6

Each page contains a list of items. We can get the items from a page by using the `get_items` method.

In [24]:
page1_items = pg.get_items(pages[0])
page1_items[0]

{'id': 'https://lux.collections.yale.edu/data/person/22e467d0-8f37-4fd8-a317-7eb096375ee0',
 'type': 'Person'}

We can get the data for an item by using the `get_item_data` method.

In [25]:
print(pg.get_item_data(page1_items[0]))

{'id': 'https://lux.collections.yale.edu/data/person/22e467d0-8f37-4fd8-a317-7eb096375ee0', 'born': {'type': 'Birth', 'timespan': {'type': 'TimeSpan', 'end_of_the_end': '1940-12-31T23:59:59', 'begin_of_the_begin': '1940-01-01T00:00:00', '_seconds_since_epoch_begin_of_the_begin': -946771200, '_seconds_since_epoch_end_of_the_end': -915148801}}, 'type': 'Person', '_label': 'Adams, J. D. R.', '@context': 'https://linked.art/ns/v1/linked-art.json', 'equivalent': [{'id': 'https://d-nb.info/gnd/138381100', 'type': 'Person', '_label': 'Adams, J. D. R.'}, {'id': 'http://id.loc.gov/authorities/names/no94009355', 'type': 'Person', '_label': 'Adams, J. D. R.'}, {'id': 'http://viaf.org/viaf/100395370', 'type': 'Person', '_label': 'Adams, J. D. R.'}, {'id': 'https://linked-art.library.yale.edu/node/5e8de466-b82b-4491-9708-f2da1d8a979e', 'type': 'Person', '_label': 'Adams, J. D. R.'}], 'classified_as': [{'id': 'https://lux.collections.yale.edu/data/concept/6f652917-4c07-4d51-8209-fcdd4f285343', 'type

Let's take a look at the label for the first item.

In [27]:
print(pg.get_item_data(page1_items[0])["_label"])

Adams, J. D. R.


### Bringing it All Together

We can now bring it all together by iterating over the pages and items. Let's iterate over the first 2 pages and print the label for each item.

In [29]:


for i, page in enumerate(pg.get_page_data_all(start_page=1, end_page=3), 1):
    print(f"Page {i}:", page["id"])
    for j, item in enumerate(pg.get_items(page)):
        print(f"Item {j}:", pg.get_item_data(item)["_label"])

Page 1: https://lux.collections.yale.edu/api/search/agent?q=%7B%22AND%22%3A%5B%7B%22name%22%3A%22John%20Adams%22%7D%5D%7D&page=2&pageLength=20
Item 0: Adams, J. D. R.
Item 1: Adams, John W., 1956-
Item 2: Lowe, John Adams, 1881-
Item 3: Adams, John Quincy, 1891-
Item 4: Samuel and John Adams (Firm)
Item 5: Henry, John A. (John Adams)
Item 6: Adams, John Stokes, 1864-1954
Item 7: Adams, John R., 1900-1994
Item 8: Adams, John Clarke, 1910-2000
Item 9: Adams, John, 1961-
Item 10: Adams, John B. (John Ballard), 1836-
Item 11: Adams, D. J. (David John)
Item 12: Adams, J. B.
Item 13: Adams, J. A.
Item 14: Adams, John, -1866
Item 15: Adams, John, 1857-1934
Item 16: Adams, John, Esq. of Waltham Abbey
Item 17: Brendon, J. A. (John Adams), 1884-
Item 18: Adams, John W. (John Wagstaff), 1915-1995
Item 19: Adams, John Crawford
Page 2: https://lux.collections.yale.edu/api/search/agent?q=%7B%22AND%22%3A%5B%7B%22name%22%3A%22John%20Adams%22%7D%5D%7D&page=3&pageLength=20
Item 0: McLaughlin, John A. (J